### PR TITLE
Add model loading abstraction

### PR DESCRIPTION
**Pull request by regadas**
Proposal to make #29 possible.

### CODE REVIEW

File: zoltar-core/src/main/java/com/spotify/zoltar/loaders/Memoizer.java [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+  /**
+   * Creates a memoized model loader.
+   *
+   * @param loader ModelLoader to be memoized.
+   * @param <M> Model instance type.
+   * @return Memoized loader.
+   */
+  static <M extends Model<?>> Memoizer<M> memoize(final ModelLoader<M> loader) {
+    final AtomicReference<CompletionStage<M>> value = new AtomicReference<>();
```
Comment (ravwojdyla) [inline zoltar-core/src/main/java/com/spotify/zoltar/loaders/Memoizer.java]: You might want to add comment here why you:
 * declare the AtomicReference here
 * why you use double locking

File: zoltar-core/src/main/java/com/spotify/zoltar/loaders/Memoizer.java [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * Memoizes the result of the supplied {@link ModelLoader}.
+ *
+ * @param <M> Model instance type.
+ */
+@FunctionalInterface
+public interface Memoizer<M extends Model<?>> extends ModelLoader<M> {
```
Comment (ravwojdyla) [inline zoltar-core/src/main/java/com/spotify/zoltar/loaders/Memoizer.java]: `Memoizer` sounds too generic, how about `ModelMemoizer` or `MemoizableModel`.

File: zoltar-models/src/main/java/com/spotify/zoltar/ModelLoaders.java [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+import javax.annotation.Nullable;
+import org.tensorflow.Graph;
+import org.tensorflow.framework.ConfigProto;
+
+/**
+ * This class consists exclusively of static methods that return Model Loaders.
+ *
+ * <p>This is the public entry point for get get a Model.</p>
+ */
+public final class ModelLoaders {
```
Comment (ravwojdyla) [inline zoltar-models/src/main/java/com/spotify/zoltar/ModelLoaders.java]: This I guess deprecates `Models` right, could you make make it private if still used internally or remove if obsolete?

File: zoltar-tensorflow/src/main/java/com/spotify/zoltar/tf/TensorFlowGraphLoader.java [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+   *                 local filesystem, resource, GCS etc.
+   * @param config   optional TensorFlow {@link ConfigProto} config.
+   * @param prefix   optional prefix that will be prepended to names in the graph.
+   */
+  static TensorFlowGraphLoader create(final String modelUri,
+                                      @Nullable final ConfigProto config,
+                                      @Nullable final String prefix) {
+    final ModelLoader<TensorFlowGraphModel> loader = ModelLoader
+        .lift(() -> TensorFlowGraphModel.create(URI.create(modelUri), config, prefix))
+        .with(Memoizer::memoize)
```
Comment (ravwojdyla) [inline zoltar-tensorflow/src/main/java/com/spotify/zoltar/tf/TensorFlowGraphLoader.java]: Maybe `Memoizer` and `Preloader` should be accessable via `Models` or `Loaders`, like:

```
ModelLoaders.Memoizer
ModelLoaders. Preloader
```
making `ModelLoaders` de facto entry point to the model API?

File: zoltar-models/src/main/java/com/spotify/zoltar/ModelLoaders.java (hunk 3 above)
Comment (regadas) [reply zoltar-models/src/main/java/com/spotify/zoltar/ModelLoaders.java]: 👍 removing it.

**Pull request merged**